src/core/Carousel/Carousel.tsx

```diff
@@ -0,0 +1,298 @@
+          onClick={() =>
+            setCurrentIndex((old) => (slideCount + old + 1) % slideCount)
+          }
+          data-pressed={keysPressed['ArrowRight'] || undefined}
+        >
+          <SvgChevronRight />
+        </IconButton>
+      </div>
+    </nav>
+  );
+};
+
+const CarouselDots = (
+  props: {
+    length?: number;
+    maxCount?: number;
+  } & React.ComponentPropsWithoutRef<'ol'>,
+) => {
+  const { maxCount = 5, className, children, ...rest } = props;
+  const context = React.useContext(CarouselContext);
+  if (!context) {
+    throw new Error('TODO: this needs to be usable without context');
+  }
+
+  const { slideCount, currentIndex, setCurrentIndex } = context;
+
+  useTheme();
+
+  const dots = React.useMemo(
+    () =>
+      Array(slideCount)
+        .fill(null)
+        .map((_, index) => {
+          const firstDotIndex = Math.max(
+            0,
+            Math.min(
+              currentIndex - Math.ceil(maxCount / 2) + 1,
+              slideCount - maxCount,
+            ),
+          );
+          const lastDotIndex = Math.min(
+            slideCount - 1,
+            Math.max(currentIndex + Math.floor(maxCount / 2), maxCount - 1),
+          );
+
+          const isFirstSmallDot =
+            (index === firstDotIndex && index !== 0) ||
+            (index === lastDotIndex && index !== slideCount - 1);
+          const isSecondSmallDot =
+            (index === firstDotIndex + 1 && index !== 1) ||
+            (index === lastDotIndex - 1 && index !== slideCount - 2);
+
+          return (
+            <li key={index}>
+              <CarouselDot
+                aria-label={`Slide ${index}`}
+                isActive={index === currentIndex}
+                onClick={() => setCurrentIndex(index)}
+                isFirst={isFirstSmallDot}
+                isSecond={isSecondSmallDot}
+                isInvisible={index < firstDotIndex || index > lastDotIndex}
+              />
+            </li>
+          );
+        }),
+    [slideCount, maxCount, currentIndex, setCurrentIndex],
+  );
+
+  return (
+    <ol
+      className={cx('iui-carousel-navigation-dots', className)}
+      role='tablist'
+      {...rest}
+    >
+      {children ?? dots}
+    </ol>
+  );
+};
+
+const CarouselDot = (
+  props: {
+    isActive?: boolean;
+    isFirst?: boolean;
+    isSecond?: boolean;
+    isInvisible?: boolean;
+  } & React.ComponentPropsWithoutRef<'button'>,
+) => {
+  const {
+    isActive,
+    isFirst,
+    isSecond,
+    isInvisible,
+    className,
+    ...rest
+  } = props;
+
+  return (
+    <button
+      type='button'
+      role='tab'
+      tabIndex={-1}
+      className={cx(
+        'iui-carousel-navigation-dot',
+        {
+          'iui-active': isActive,
+          'iui-first': isFirst,
+          'iui-second': isSecond,
+          'iui-invisible': isInvisible,
+        },
+        className,
+      )}
+      {...rest}
+    />
+  );
+};
+
+Carousel.Slider = CarouselSlider;
```

Comment:
Would you move then those compnents in separate files or leave all in here?

Reply:
Haven't decided yet. It's kinda neat that these are "internal" components only exposed through `Carousel`

Reply:
I think we can leave that `Carousel` has subcomponents as its properties but we should have each component in separate file - for me it is more readable.

Reply:
Would it make sense to have CarouselSlider and CarouselSlide in one file and CarouselDots and CarouselDot in one file? They are very tightly coupled

Reply:
ok, split into 7 files then.

Tests currently pending

Reply:
a maaaan...
Could have fixed everything and then just move those files, if so important... Now need to review everything in not very comfortable way :./

Reply:
what's not comfortable? 

I made only two commits, the first one is purely cosmetic: moving stuff around, adding comments, creating types for props, destructuring props, etc.
![image](https://user-images.githubusercontent.com/9084735/158812076-ae11a2e3-bf46-4ddf-96de-4f373ee2da6d.png)